### PR TITLE
[Impeller] round scale for glyph atlas cache to 2 decimal places.

### PR DESCRIPTION
**Pull request by jonahwilliams**
Fixes https://github.com/flutter/flutter/issues/130750

### CODE REVIEW

**Comment by jonahwilliams** [comment]
Maybe this is OK? 🤷‍♂️

File: impeller/entity/contents/text_contents.cc [hunk 1]
```diff
@@ -25,6 +25,11 @@ TextContents::TextContents() = default;
 
 TextContents::~TextContents() = default;
 
+// static
+Scalar TextContents::RoundFontScale(Scalar value) {
+  return std::round(value * 10.0) / 10.0;
```
Comment (bdero) [inline impeller/entity/contents/text_contents.cc:30]: I think this snapping would need to be done logarithmically (apply log, apply snapping, invert with exponential)? Also, the snapping precision would need to somehow incorporate the text frame's font size as well.

For example, the difference between 0.1 and 0.2 is potentially huge, getting worse as the given font size increases.

**Review by bdero — COMMENTED** [review]
I haven't tried it out, but something like this should be OK

File: impeller/entity/contents/text_contents.cc (hunk 1 above)
Comment (jonahwilliams) [reply impeller/entity/contents/text_contents.cc:30]: hmm thats a good point. I'll need to investigate this a bit more.

Comment (jonahwilliams) [reply impeller/entity/contents/text_contents.cc]: okay, after much thinking I've come up with an OK solution.... round to one decimal if the point size is < 25, Round to two decimals if > 50.  🤷‍♂️

Comment (bdero) [reply impeller/entity/contents/text_contents.cc]: I don't think this covers all cases, would something like this work?

```cpp
// The point_size is multiplied into the total scale of the font later,
// and so we need to linearly increase the step precision as the point size increases.
auto precision = 20 * point_size; // The constant here may need to be adjusted for best perf.
return std::pow(kE, std::round(std::log(scale) * precision) / precision);
```

image

https://www.desmos.com/calculator/sqpcws9bfv

File: impeller/typographer/text_frame.cc [hunk 2]
```diff
@@ -79,13 +79,23 @@ bool TextFrame::MaybeHasOverlapping() const {
   return false;
 }
 
+// static
+Scalar TextFrame::RoundScaledFontSize(Scalar scale, Scalar point_size) {
+  if (point_size <= 25) {
+    return std::round(scale * 10) / 10;
+  }
+  return std::round(scale * 100) / 100;
+}
```
Comment (bdero) [inline impeller/typographer/text_frame.cc:88]: I don't think this covers all cases, would something like this work?

```cpp
// The point_size is multiplied into the total scale of the font later,
// and so we need to linearly increase the step precision as the point size increases.
auto precision = 20 * point_size; // The constant here may need to be adjusted for best perf.
return std::pow(kE, std::round(std::log(scale) * precision) / precision);
```

image

https://www.desmos.com/calculator/sqpcws9bfv

Comment (jonahwilliams) [reply impeller/typographer/text_frame.cc:88]: This curve is going to introduce a lot of churn for values that are already fairly stable.

For the default text scale of 3.5x (the DPR) we can end up converting this to  3.49908 or  3.50127 based on the exact text scale. I don't we're going to be able to use an log based rounding.

**Comment by jonahwilliams** [comment]
Running though flutter gallery, I'm looking at how much error we introduce by rounding. For example on one page we round 3.04602 -> 3.

```
E/flutter (21011): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.414219, 0.38833)
E/flutter (21011): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0, 0)
E/flutter (21011): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.215739, 0.414219)
E/flutter (21011): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.258887, 0.310664)
E/flutter (21011): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.258887, 0.310664)
E/flutter (21011): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.293405, 0.414219)
E/flutter (21011): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0862956, 0.422848)
E/flutter (21011): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.310664, 0.310664)
E/flutter (21011): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.293405, 0.302035)
E/flutter (21011): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.258887, 0.310664)
E/flutter (21011): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.310664, 0.422848)
E/flutter (21011): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0604069, 0.414219)
E/flutter (21011): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.302035, 0.310664)
E/flutter (21011): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0, 0)
E/flutter (21011): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.172591, 0.302035)
E/flutter (21011): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.302035, 0.310664)
E/flutter (21011): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.215739, 0.371071)
E/flutter (21011): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.258887, 0.310664)
E/flutter (21011): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0862956, 0.422848)
E/flutter (21011): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0604069, 0.414219)
E/flutter (21011): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0, 0)
E/flutter (21011): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.258887, 0.310664)
E/flutter (21011): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.310664, 0.414219)
E/flutter (21011): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.310664, 0.414219)
```

That results in almost half a pixel difference in some cases, though others are much smaller.

**Comment by jonahwilliams** [comment]
If I round to two decimal places, that gives way less error, naturally:

```
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(96)] 3.04602->3.05
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0288233, 0.053671)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0258416, 0.0516832)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0198781, 0.0377685)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.010933, 0.0506893)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0258416, 0.0377685)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0248477, 0.0387624)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(96)] 3.04602->3.05
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0357807, 0.0335444)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0, 0)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0186358, 0.0357807)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0223629, 0.0268355)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0223629, 0.0268355)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0253446, 0.0357807)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.00745431, 0.0365261)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0268355, 0.0268355)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0253446, 0.0260901)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0223629, 0.0268355)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0268355, 0.0365261)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.00521801, 0.0357807)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0260901, 0.0268355)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0, 0)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0149086, 0.0260901)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0260901, 0.0268355)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0186358, 0.0320535)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0223629, 0.0268355)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.00745431, 0.0365261)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.00521801, 0.0357807)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0, 0)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0223629, 0.0268355)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0268355, 0.0357807)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0268355, 0.0357807)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(96)] 3.5->3.5
```

**Comment by jonahwilliams** [comment]
If I blow up the text size to the largest possible supported system size, the errors stay pretty OK looking at 2 decimal places

```
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(96)] 3.39525->3.4
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.128199, 0.120186)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0, 0)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0667702, 0.128199)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0801243, 0.0961491)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0801243, 0.0961491)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0908075, 0.128199)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0267081, 0.13087)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0961491, 0.0961491)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0908075, 0.0934783)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0801243, 0.0961491)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0961491, 0.13087)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0186957, 0.128199)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0934783, 0.0961491)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(96)] 3.39525->3.4
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0534162, 0.0934783)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0934783, 0.0961491)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0667702, 0.114845)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0801243, 0.0961491)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0267081, 0.13087)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0186957, 0.128199)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0, 0)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0801243, 0.0961491)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0961491, 0.128199)
E/flutter (21427): [ERROR:flutter/impeller/typographer/text_frame.cc(101)] (0.0961491, 0.128199)
```

**Comment by jonahwilliams** [comment]
2 decimal places of rounding is also sufficient to prevent the glyph atlas from filling up by scrolling through the cards on flutter/gallery.

**Review by bdero — APPROVED** [review]
If it seems like it won't introduce fidelity issues for reasonable scale/point_size inputs, I say let's go for it.

Maybe include your error calculation/debug print code, but commented out? That may come in handy later.

**Pull request merged**